ability to define paginator accordingly to documentation

pxpm opened the pull request:
refs: #3326 

problem: at the moment the only way to define the page length is using a multi dimensional array either:
```php
setPageLengthMenu([[100, 200, 300]]);
//or
setPageLengthMenu([[100, 200, 300],['label1', 'label2', 'label3']]);
```

In the documentation: https://backpackforlaravel.com/docs/4.1/crud-cheat-sheet#page-length the suggested way is: `setPageLengthMenu([100, 200, 300]);` and that will fail. 

Instead of just changing the docs, I think that the way that docs describes is the way I see most developers wanting to edit their paginator, just provide custom options to select as a simple array. 

this PR should now allow, with backward compatibility:

```php
setPageLengthMenu([100, 200, 300]);
setPageLengthMenu(100);
setPageLengthMenu([100 => 'a', 200 => 'b', 300 => 'c']);
```

Also added tests for this scenarios

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/Laravel-Backpack/CRUD/inspections/3c634563-b701-4f7a-8b8d-cb8caef5816e) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/Laravel-Backpack/CRUD/inspections/3c634563-b701-4f7a-8b8d-cb8caef5816e) completed: **1 updated code elements**

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/Laravel-Backpack/CRUD/inspections/3b4e9be8-f33e-41d0-97c3-796d94fa26ac) was created.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/Laravel-Backpack/CRUD/inspections/e00c1c37-ac0b-437c-8442-3b7c025b9bae) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/Laravel-Backpack/CRUD/inspections/3b4e9be8-f33e-41d0-97c3-796d94fa26ac) completed: **2 new issues, 8 updated code elements**

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/Laravel-Backpack/CRUD/inspections/e00c1c37-ac0b-437c-8442-3b7c025b9bae) completed: **2 new issues, 8 updated code elements**

promatik commented on src/app/Library/CrudPanel/Traits/Read.php:
```diff
@@ -225,6 +232,36 @@ public function addCustomPageLengthToPageLengthMenu()
     public function setPageLengthMenu($menu)
     {
+        // validates the correct building of the menu array
+        if (isset($menu[0]) && ! is_null($menu[0])) {
```
```suggestion
        if (isset($menu[0])) {
```

Is this validation needed? The tests run without it.

promatik reviewed (changes requested):
@pxpm I noticed this fails in a situation like;
```php
$this->crudPanel->setPageLengthMenu([0 => 'v', 30 => 't', 40 => 'q']);
```

But no one's going to use 0 for the page length right!?

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/Laravel-Backpack/CRUD/inspections/dcb60f60-c96d-4b47-af2e-0701a6f9df1b) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/Laravel-Backpack/CRUD/inspections/dcb60f60-c96d-4b47-af2e-0701a6f9df1b) completed: **2 new issues, 7 updated code elements**

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/Laravel-Backpack/CRUD/inspections/396f223a-f4a5-456d-af85-b2a36d1f8a50) was created.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/Laravel-Backpack/CRUD/inspections/53ee047d-afb3-4907-9338-cffb27246f7d) was created.

pxpm commented:
Hello @promatik 

Thank you very much for spotting that. Indeed, it does not make sound to use `0` and it did not came to my mind, nice catch. 

But while having a look at this, I remembered before we added the option to show ALL entries, using [-1 => 'All'], so I decided to push more tests and refactored the code so it would not break even with 0.

Let me know :)

Thanks again man, you rock 👯‍♂️ 

Pedro

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/Laravel-Backpack/CRUD/inspections/396f223a-f4a5-456d-af85-b2a36d1f8a50) completed: **3 new issues, 11 updated code elements**

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/Laravel-Backpack/CRUD/inspections/53ee047d-afb3-4907-9338-cffb27246f7d) completed: **3 new issues, 11 updated code elements**

pxpm commented on tests/Unit/CrudPanel/CrudPanelReadTest.php:
```diff
@@ -385,4 +387,163 @@ public function testGetRelatedEntriesAttributesFromHasOneWithAcessor()
+    /**
+     * Tests define paginator length with multi array [[20, 30, 40],['v', 't', 'q']].
+     *
+     * @group failing
```
@promatik dunno if you know about it. But I will let you know here why this `@group failing`. 

Using this `@group` technique we can run specific tests instead of the whole testsuit. 

After defining the tests you want to run in the same group just `vendor/bin/phpunit --testdox --group=failing`

promatik replied:
Great @pxpm didn't know about it, I was using filter to run tests related to this PR.
`vendor/bin/phpunit --testdox --filter CrudPanelChangePaginatorLength`. It acts like a regex match, so it will match all the tests with that string.

promatik commented on src/app/Library/CrudPanel/Traits/Read.php:
```diff
@@ -203,17 +203,24 @@ public function getDefaultPageLength()
-                if (is_array($page_length_choices)) {
-                    array_unshift($page_length_choices, $this->getDefaultPageLength());
-                }
+        $values = $this->getOperationSetting('pageLengthMenu')[0];
+        $labels = $this->getOperationSetting('pageLengthMenu')[1];
+
+        // This is a condition that should be always true.
+        if (is_array($values) && is_array($labels)) {
+            $position = array_search($this->getDefaultPageLength(), $values);
+            if ($position !== false) {
+                $defaultLabel = isset($labels[$position]) ? $labels[$position] : $this->getDefaultPageLength();
+                array_unshift($values, $this->getDefaultPageLength());
+                array_unshift($labels, $defaultLabel);
+
+                //now make it unique.
+                $values = array_unique($values);
+                $labels = array_unique($labels);
             }
         }
+
+        $this->setOperationSetting('pageLengthMenu', [$values, $labels]);
```
@pxpm do we need this function?

Because all the logic is inside `setPageLengthMenu` and with this function code is not working 😅

When I reverted this, it worked once again.

pxpm replied:
Setting up with 0 will fail in the database in the stock backpack search functionality, but if you override, developer could use the 0 as a placeholder for something and handle it on his side. 

Like we talked before, I don't think that a regular dev that wants to setup a paginator will use 0, but I think we should not error on '0' because developer could handle it on his side. 

On the other hand, if this 0 handling starts to get out of control I think we better add to the docs that `0` is not supported and remove everything that is related to handle `0` as a key in the array. 
**I am starting feeling that this is the way to go instead of adding that much code to handle `0`**

What do you think @promatik ?

pxpm replied:
About the function itself, that function exists because you can simply `add a new value to the backpack paginator and make it default` without the need to go copy the whole backpack paginatior and add the value manuay. 

You can just config `CRUD::setOperationSetting('defaultPageLength', 30)` in your ListOperation and we will pick it and add it to paginator making it the default.

promatik replied:
@pxpm I agree that we should avoid to add this extra layer os code/conditions to handle "0", but what should developers use for the "All"?

Regarding the function, now I know 😅, but something is not working 😬
I may later have a look once again 👍

pxpm replied:
Hello @promatik, to emulated the `all` we used `-1` before, that's why I also added tests for -1. 

I am going to have a talk with @tabacitu about this, I think we should remove the extra coded I added to support 0, and just let devs known that `0` is a non-valid paginator number, and to use -1 if they want to show all.

I will come back to you asap 👍 thanks again @promatik

pxpm replied:
Like we discussed, let's not support 0, there is no sense in supporting it.

promatik reviewed (changes requested):
Something that's also not working is `CRUD::setPageLengthMenu(0)` it will really use 0 to set the offset of the query, leading to an sql error.

tabacitu commented:
Todo:
- [ ] create a PR for the docs to document all types of parameters

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/Laravel-Backpack/CRUD/inspections/0bcccdb8-9172-4826-9abb-0c6b875d8014) was created.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/Laravel-Backpack/CRUD/inspections/151a8c79-4ba2-4512-b457-f1542a466c85) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/Laravel-Backpack/CRUD/inspections/0bcccdb8-9172-4826-9abb-0c6b875d8014) completed: **12 new issues, 11 updated code elements**

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/Laravel-Backpack/CRUD/inspections/151a8c79-4ba2-4512-b457-f1542a466c85) completed: **12 new issues, 11 updated code elements**

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/Laravel-Backpack/CRUD/inspections/e88d5ea3-59cd-4e71-a99e-d6942f8e441b) was created.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/Laravel-Backpack/CRUD/inspections/4bf84f04-5ae1-439b-9a7c-5b19bbddc23c) was created.

pxpm commented:
Link to docs: https://github.com/Laravel-Backpack/docs/pull/209

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/Laravel-Backpack/CRUD/inspections/e88d5ea3-59cd-4e71-a99e-d6942f8e441b) completed: **8 new issues, 11 updated code elements**

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/Laravel-Backpack/CRUD/inspections/4bf84f04-5ae1-439b-9a7c-5b19bbddc23c) completed: **8 new issues, 11 updated code elements**

promatik commented on src/app/Library/CrudPanel/Traits/Read.php:
```diff
@@ -203,31 +207,97 @@ public function getDefaultPageLength()
+            if (count($menu) !== count($menu, COUNT_RECURSIVE)) {
+                // developer defined as setPageLengthMenu([[50, 100, 300]]) or setPageLengthMenu([[50, 100, 300],['f','h','t']])
+                // we will apply the same labels as the values to the menu if developer didn't
+                if (! in_array(0, $menu[0])) {
+                    if (! isset($menu[1]) || ! is_array($menu[1])) {
+                        $menu[1] = $menu[0];
+                    }
+                } else {
+                    abort(500, 'You should not use 0 as a key in paginator. If you are looking for "ALL" option, use -1 instead.');
+                }
```
```suggestion
                if (in_array(0, $menu[0])) {
                    abort(500, 'You should not use 0 as a key in paginator. If you are looking for "ALL" option, use -1 instead.');
                }

                if (! isset($menu[1]) || ! is_array($menu[1])) {
                    $menu[1] = $menu[0];
                }
```

promatik commented on src/app/Library/CrudPanel/Traits/Read.php:
```diff
@@ -203,31 +207,97 @@ public function getDefaultPageLength()
+            }
+        } else {
+            // developer added only a single value setPageLengthMenu(10)
+            if ($menu !== 0) {
+                $menu = [[$menu], [$menu]];
+            } else {
+                abort(500, 'You should not use 0 as a key in paginator. If you are looking for "ALL" option, use -1 instead.');
+            }
```
```suggestion
            if ($menu === 0) {
                abort(500, 'You should not use 0 as a key in paginator. If you are looking for "ALL" option, use -1 instead.');
            }

            $menu = [[$menu], [$menu]];
```

promatik commented on src/app/Library/CrudPanel/Traits/Read.php:
```diff
@@ -203,31 +207,97 @@ public function getDefaultPageLength()
+            $values = array_keys($menu);
+            $labels = array_values($menu);
+
+            if (! in_array(0, $values)) {
+                return [$values, $labels];
+            }
+
+            abort(500, 'You should not use 0 as a key in paginator. If you are looking for "ALL" option, use -1 instead.');
```
```suggestion
            if (in_array(0, $values)) {
                abort(500, 'You should not use 0 as a key in paginator. If you are looking for "ALL" option, use -1 instead.');
            }

            return [$values, $labels];
```

promatik commented on src/app/Library/CrudPanel/Traits/Read.php:
```diff
@@ -203,31 +207,97 @@ public function getDefaultPageLength()
+        } else {
+            // developer defined length as setPageLengthMenu([50, 100, 300])
+            // we will use the same values as labels
+            if (! in_array(0, $menu)) {
+                return [$menu, $menu];
+            }
+
+            abort(500, 'You should not use 0 as a key in paginator. If you are looking for "ALL" option, use -1 instead.');
```
```suggestion
            if (in_array(0, $menu)) {
                abort(500, 'You should not use 0 as a key in paginator. If you are looking for "ALL" option, use -1 instead.');
            }

            return [$menu, $menu];
```

pxpm replied:
I am not sure we should validate for 0 before the actual code. 

I think we should build it as `the most expected scenario` to be at first. And it's totally unexpected that someone will use 0. 

@tabacitu can you help us to decide here ?

tabacitu replied:
Gladly buuuut... I don't understand the issue sorry 😂 Both your original code and @promatik 's suggestion look fine to me - the way I understand it is we try to see if `0` is an array key - if so we throw a nasty error to let the dev know he shouldn't do that.

I'd prefer returning early like @promatik proposed.

tabacitu replied:
Oh yeah now I get it 😂 The question is whether to:
- (a) assume it doesn't have zero
- (b) assume it might have zero

I get what you're saying now @pxpm . There's small performance difference there, but for the sake of following the "return early" principle, we can use @promatik 's suggestion this time. It's easier to read the functions if you assume their structure is exceptions first, then at the very end what happens in the best case scenario (the end-game, if you will).

promatik replied:
I think in terms of performance I believe it is the same, because we are also validation it is **different** from zero, or zero is **not** in array.

In my perspective it is easier to read the code in this case when you do "nasty" code at the beginning;
_If there's a zero abort!!! Ok then, do everything that should be done ..._

And in the future, if we ever come to realize we should add zero support, or zero safe fallback, or even remove this abort for whatever reason, it's easy to just remove this lines, to commit and to review it also.

But again, this is an opinionated suggestion 😅

promatik reviewed (changes requested):
This is an opinionated review, I think it's more expressive and readable to have the validations this away. What do you think @pxpm?

promatik commented:
@pxpm what do you think about creating a function to avoid repeating the `0` validation and `abort(500)` so many times?
Something like;

```php
private function abortOnPageLengthZero($value)
{
    if ($value === 0 || (is_array($value) && in_array(0, $value))) {
        abort(500, 'You should not use 0 as a key in paginator. If you are looking for "ALL" option, use -1 instead.');
    }
}
```

For instance, for `setDefaultPageLength` instead of;

```php
public function setDefaultPageLength($value)
{
    if ($value !== 0) {
        $this->setOperationSetting('defaultPageLength', $value);
    } else {
        abort(500, 'You should not use 0 as a key in paginator. If you are looking for "ALL" option, use -1 instead.');
    }
}
```

we could have;

```php
public function setDefaultPageLength($value)
{
    // validate page length
    $this->abortOnPageLengthZero($value);

    $this->setOperationSetting('defaultPageLength', $value);
}
```

Don't know about the name, it could be `abortOnPageLengthZero` or `validateNonZeroPageLength` or just `validatePageLength` ...

promatik reviewed (changes requested):
@pxpm `getPageLengthMenu` also needs to be changed. As it is right now it's not supporting the current multiplicity of possible values.

In this branch in the Monster CRUD this is the result:
![image](https://user-images.githubusercontent.com/1838187/101260440-c4351f00-3727-11eb-93e2-e0378e825363.png)

Also, it's translating the values, only in case there is a second array, so it doesn't support a system where `[10 => 10, -1 => 'backpack::crud.all']`.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/Laravel-Backpack/CRUD/inspections/2cb95eca-3fee-48c3-a7c5-46c9865379f1) was created.

pxpm commented:
> @pxpm `getPageLengthMenu` also needs to be changed. As it is right now it's not supporting the current multiplicity of possible values.
> 
> In this branch in the Monster CRUD this is the result:
> ![image](https://user-images.githubusercontent.com/1838187/101260440-c4351f00-3727-11eb-93e2-e0378e825363.png)
> 
> Also, it's translating the values, only in case there is a second array, so it doesn't support a system where `[10 => 10, -1 => 'backpack::crud.all']`.

Hello @promatik the problem you referencend only happens if you add a `defaultPageLength` that is already on the `paginator array`, when we move it to be the first in array, we endup with array like: [0 => 10, 2 => 20, 3 => 30]. Notice is missing key "1", when json_encoding it will fail and show like you mentioned. 

I'v fixed this by running the array through `array_values` that will reset the array keys.

About the translation, it will run `trans()` not matter how you add your paginator because we ALLWAYS convert it to a multi-dimensional array `[[a1],[a2]]`

Please take in mind that datatables have some kind of `internal cache`, and you should clear localStorage when testing this out, otherwise it might show you inacurate data after a change in the controller.

Best,
Pedro

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/Laravel-Backpack/CRUD/inspections/2cb95eca-3fee-48c3-a7c5-46c9865379f1) completed: **8 new issues, 11 updated code elements**

tabacitu commented on src/app/Library/CrudPanel/Traits/Read.php:
```diff
@@ -184,7 +184,11 @@ public function disableBulkActions()
     public function setDefaultPageLength($value)
     {
-        $this->setOperationSetting('defaultPageLength', $value);
+        if ($value !== 0) {
+            $this->setOperationSetting('defaultPageLength', $value);
+        } else {
+            abort(500, 'You should not use 0 as a key in paginator. If you are looking for "ALL" option, use -1 instead.');
+        }
```
I agree as a principle we should try to do this more often - "return early". If we can do an `if` instead of `if else` we should do it. Good catch @promatik 👍

tabacitu commented:
> Please take in mind that datatables have some kind of internal cache, and you should clear localStorage when testing this out, otherwise it might show you inacurate data after a change in the controller.

@pxpm will this caching make this a breaking change? 👀

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/Laravel-Backpack/CRUD/inspections/254d5325-5e21-4b3b-8401-d09bfab777b1) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/Laravel-Backpack/CRUD/inspections/254d5325-5e21-4b3b-8401-d09bfab777b1) completed: **8 new issues, 11 updated code elements**

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/Laravel-Backpack/CRUD/inspections/af617357-459e-466c-b0ca-b8e26822fbaa) was created.

pxpm commented:
@tabacitu I do not think it is a BC. 

What developer have defined now will work the same. We might add this to the docs to be safe, you will only notice any difference if you change paginator.

default backpack:
![image](https://user-images.githubusercontent.com/7188159/101618978-8cb9c180-3a0a-11eb-8264-8c86adb8ab7a.png)

now changing: 

```php
$this->crud->setPageLengthMenu([[50, 100, 300],['backpack::crud.all','f','v']]);

//notice that the `defaultPageLength` still backpack default 10. 
```

![image](https://user-images.githubusercontent.com/7188159/101619171-ce4a6c80-3a0a-11eb-9f82-af2202785209.png)

```php
//so far so good. now let's get rid of default 10, and make 50 the default.

$this->crud->setDefaultPageLength(50);
```

![image](https://user-images.githubusercontent.com/7188159/101619374-179abc00-3a0b-11eb-8290-1a44b015c7b5.png)

That 10 is still "cached" in localStorage. but we don't have it in selectable options, we had it before because it was the default page length.

If you select ANY options of the available options it will reset this cache and will behave fine. 

So this problem will apply to everyone that used the crud table atleast one time (will have the cache in LS), and then changes the paginator in crud. 

The only way I can think to automate this process is to check if the current `defaultPageLength` equals the one in LS, if not, reset the LS. I'v just updated this PR with that check. https://github.com/Laravel-Backpack/CRUD/pull/3332/commits/ae3018ad86e54450725d0b5584c94bf47ffcfb2e

Let me know,
Pedro

promatik commented:
@pxpm have you seen this comment? https://github.com/Laravel-Backpack/CRUD/pull/3332#issuecomment-739326211
Do you think it's worth?

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/Laravel-Backpack/CRUD/inspections/af617357-459e-466c-b0ca-b8e26822fbaa) completed: **8 new issues, 11 updated code elements**

pxpm commented:
> @pxpm have you seen this comment? [#3332 (comment)](https://github.com/Laravel-Backpack/CRUD/pull/3332#issuecomment-739326211)
> Do you think it's worth?

Sorry, must have slipped me. 

100% agree in refactoring to a new function, I will also update the tests.

tabacitu commented:
Let's do it! Let's create `abortIfInvalidPageLength($value)`

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/Laravel-Backpack/CRUD/inspections/a30fe6b7-66b2-4307-bfd3-d1faa6834dbe) was created.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/Laravel-Backpack/CRUD/inspections/00ba6aec-0f9e-4fa1-b998-b5ddde5c6dde) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/Laravel-Backpack/CRUD/inspections/00ba6aec-0f9e-4fa1-b998-b5ddde5c6dde) completed: **11 new issues, 11 updated code elements**

promatik reviewed (changes requested):
@pxpm I saw in the code you are on purpose moving the default page lenght to the begging of the stack, it can make a dropdown look like this;

![image](https://user-images.githubusercontent.com/1838187/101965332-32884e80-3c0c-11eb-8611-6534760f0ba0.png)

I set;
```php
$this->crud->setDefaultPageLength(20);
$this->crud->setPageLengthMenu([10 => 10, 20 => 20, 30 => 30, -1 => 'All']);
```

I don't know if we should be moving the default page length to the beginning, what do you think?

pxpm commented:
> @pxpm I saw in the code you are on purpose moving the default page lenght to the begging of the stack, it can make a dropdown look like this;
> 
> ![image](https://user-images.githubusercontent.com/1838187/101965332-32884e80-3c0c-11eb-8611-6534760f0ba0.png)
> 
> I set;
> 
> ```
> $this->crud->setDefaultPageLength(20);
> $this->crud->setPageLengthMenu([10 => 10, 20 => 20, 30 => 30, -1 => 'All']);
> ```
> 
> I don't know if we should be moving the default page length to the beginning, what do you think?

Indeed @promatik 

That's what we were doing before, and I kept the same behaviour. Unfortunatelly I don't think we can change it or it will be even weirder. 

You get 20 results in the table, and the pagination is showing `10` selected. 

Let me know if you are aware of a solution for this.

Best,
Pedro

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/Laravel-Backpack/CRUD/inspections/607a3d4f-fb48-4719-aa8d-6d4298a5627c) was created.

tabacitu commented:
I agree with both of you: 
- having 20, 10, 30 is weird;
- since it was the previous behaviour, let's leave it like this for now and be done with it (20-10-30);

@promatik could you please be the green-lighter here and:
- [ ] double-check that everything's working fine (except for that) if we merge this PR as-is;
- [ ] create a new issue to solve the ordering in 4.2 (problem for future us);

Let's get this merged!

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/Laravel-Backpack/CRUD/inspections/607a3d4f-fb48-4719-aa8d-6d4298a5627c) completed: **8 new issues, 11 updated code elements**

promatik commented on src/app/Library/CrudPanel/Traits/Read.php:
```diff
@@ -203,31 +205,89 @@ public function getDefaultPageLength()
+            // if position is not false we already have that value in the pagination array
+            // we are just going to make it first element in array
+            if ($position !== false) {
+                array_unshift($values, $this->getDefaultPageLength());
+                array_unshift($labels, $labels[$position]);
+            } else {
+                // if it's not in array we add it as the first element
+                array_unshift($values, $this->getDefaultPageLength());
+                array_unshift($labels, $this->getDefaultPageLength());
```
@pxpm can you try this?
With my tests it solves the problem of positioning the default page length on the array.

```suggestion
                for ($i = 0; $i < count($values); $i++) {
                    if ($values[$i] > $this->getDefaultPageLength() || $values[$i] === -1) {
                        array_splice($values, $i, 0, $this->getDefaultPageLength());
                        array_splice($labels, $i, 0, $this->getDefaultPageLength());
                        break;
                    }
                    if ($i === count($values) - 1) {
                        $values[] = $this->getDefaultPageLength();
                        $labels[] = $this->getDefaultPageLength();
                        break;
                    }
                }
```

![image](https://user-images.githubusercontent.com/1838187/103368241-fe8fa300-4abe-11eb-8b29-8c16988dc3f4.png)

![image](https://user-images.githubusercontent.com/1838187/103368277-1109dc80-4abf-11eb-9a9d-d6d538c2d394.png)

promatik replied:
As far as I understood, it is properly selecting the default page length on the dropdown menu. I just did the part of inserting 
Maybe I'm missing something? 😅

Anyway, everything is working 👌
Tested with all the possible formats.

pxpm replied:
You didn't. 

Everything works as expected. 

Just pushed your fix along with fixed tests for it. 

Thanks again \o

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/Laravel-Backpack/CRUD/inspections/a02b17a5-01c8-4f0c-84bd-ce092caeeed7) was created.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/Laravel-Backpack/CRUD/inspections/ccb5739d-c6a1-4138-8b8f-a60768a0dc30) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/Laravel-Backpack/CRUD/inspections/ccb5739d-c6a1-4138-8b8f-a60768a0dc30) completed: **9 new issues, 12 updated code elements**

promatik commented on src/app/Library/CrudPanel/Traits/Read.php:
```diff
@@ -203,31 +205,87 @@ public function getDefaultPageLength()
+        $values = $this->getOperationSetting('pageLengthMenu')[0];
+        $labels = $this->getOperationSetting('pageLengthMenu')[1];
+        // this is a condition that should be always true.
+        if (is_array($values) && is_array($labels)) {
```
```suggestion
        if (array_search($this->getDefaultPageLength(), $values) === false) {
```
@pxpm this is missing on the new setup. We need to validate the value doesn't exist.

Another thing, I removed here the validations to check if both values are array, for my observation, this values come here as being always arrays. Correct me if I'm wrong.

pxpm replied:
Hey @promatik are you sure that we need to validate that the value doesn't exist ? 

It works in tests without it when adding a value that does not exist: 

```php
private $defaultPaginator = [[10, 20, 30], ['t1', 't2', 't3']];

    /**
     * Tests if table paginator adds default option non-existent at time in the paginator.~.
     */
    public function testCrudPanelPaginatorAddsDefaultOptionNonExistent()
    {
        $this->crudPanel->setModel(User::class);
        $this->crudPanel->setOperation('list');
        $this->crudPanel->setDefaultPageLength(40);
        $this->crudPanel->setPageLengthMenu($this->defaultPaginator);

        $this->assertCount(2, $this->crudPanel->getPageLengthMenu());
        $this->assertTrue(in_array(40, $this->crudPanel->getOperationSetting('pageLengthMenu')[0]));
        $this->assertEquals(array_values($this->crudPanel->getPageLengthMenu()[0])[3], 40);
    }
```

Technically it should be always array. I just wouldn't mind to have that extra check there.

pxpm replied:
Just ignore me. 

You are right, I was talking about wrong test. Sorry.

Fixed it. Thanks again \o

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/Laravel-Backpack/CRUD/inspections/8232cd40-51ce-4b99-adb6-d170fb5f1cc8) was created.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/Laravel-Backpack/CRUD/inspections/cbadabb7-dc7c-4bec-940d-96797720797e) was created.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/Laravel-Backpack/CRUD/inspections/4b67ef64-d2e9-4451-8d0e-8d37f15aa4b8) was created.

pxpm commented:
I just gave this one more time, added a few more checks in tests. Tested & passed. 

I think this one is ready for merge, also @tabacitu comment does not apply anymore because the issue described is already fixed here. 

Best,
Pedro

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/Laravel-Backpack/CRUD/inspections/4b67ef64-d2e9-4451-8d0e-8d37f15aa4b8) completed: **9 new issues, 11 updated code elements**

tabacitu commented:
Merged! Thanks a lot for the great work you guys. This is a relatively big change to an operation that EVERYBODY uses, it gives me a lot of confidence to merge seeing how much you've gone back&forth, that it has unit tests, and everything.

Great work! Tested & merged. Will be available in the next patch update (most likely later today).

Cheers!
